client/src/lib.rs

```diff
@@ -299,6 +301,11 @@ impl<C> ConfiguredClient<C> {
     pub fn capabilities(&self) -> Option<&get_system_info_response::Capabilities> {
         self.capabilities.as_ref()
     }
+
+    /// Returns a cloned reference to a registry with workers using this client instance
+    pub fn workers(&self) -> Arc<RwLock<SlotManager>> {
```

Comment:
I wonder if RwLock should be an implementation detail of the `SlotManager` instead of exposed here

Reply:
I think RwLock makes sense here because the majority of accesses in the critical path are just readers, it is only when a worker starts or shutdown that we need a write lock...

Reply:
But the caller shouldn't care when the implementation of `SlotManager` needs a lock. `SlotManager` impl should obtain the lock when it needs.

Reply:
In other languages I completely agree with you that you want to hide the mutex, but what I'm seeing in Rust is to explicitly use Arc<Mutex> to protect top level references to shared objects, I guess because the compiler keeps thing safe, and the mutex always owns the protected object. For example, in the same ConfiguredClient struct that contains SlotManager the headers field is  Arc<RwLock<HashMap<String, String>>>...

One of the things that worry me about Rust is that because "it is safe to share",  sometimes safe programs are much harder to understand because shared state is everywhere, and with few abstractions... But I guess that is another discussion, and I have an Erlang bias here...

Reply:
What is a top level reference? Why does core not make the user have a `Arc<RwLock<ConfiguredClient>>` instead of putting a lock around that headers field and hiding it from users? You should do the same. It is a small semantic difference, but callers can't possibly understand the internal expense of the try-reserve call and therefore can't be confident in knowing that they aren't deadlocking the system by taking a lock when calling it. Only the implementation of the `SlotManager` knows its implementation is acceptable to lock around.